std/digest/murmurhash.d

```diff
@@ -0,0 +1,860 @@
+    MurmurHash3_32_opt32 hasher;
+    hasher.put(data1);
+    hasher.put(data2);
+    // The call to 'finish' ensures:
+    // - the remaining bits are processed
+    // - the hash gets finalized
+    auto hashed = hasher.finish();
+}
+
+///
+unittest
+{
+    // Using FastMurmurHash3_32_opt32, FastMurmurHash3_128_opt32 and
+    // FastMurmurHash3_128_opt64 you gain full control over which part of the
+    // algorithm to run.
+    // This allows for maximum throughput but needs extra care.
+
+    // Data type must be the same as the hasher's element type:
+    // - uint for FastMurmurHash3_32_opt32
+    // - ulong[2] for FastMurmurHash3_128_opt32 and FastMurmurHash3_128_opt64
+    const(uint)[] data = [1, 2, 3, 4];
+    // Note the hasher starts with 'Fast'.
+    FastMurmurHash3_32_opt32 hasher;
+    // Push as many array of elements as you need. The less calls the better.
+    hasher.putBlocks(data);
+    // Put remainder bytes if needed. This method can be called only once.
+    hasher.putRemainder(ubyte(1), ubyte(1), ubyte(1));
+    // Call finalize to incorporate data length in the hash.
+    hasher.finalize();
+    // Finally get the hashed value.
+    auto hashed = hasher.getBytes();
+}
+
+public import std.digest.digest;
+
+@safe:
+
+/// MurmurHash3_32_opt32 $(D std.digest.digest) Template API.
+alias MurmurHash3_32_opt32 = Piecewise!FastMurmurHash3_32_opt32;
+/// MurmurHash3_128_opt32 $(D std.digest.digest) Template API.
+alias MurmurHash3_128_opt32 = Piecewise!FastMurmurHash3_128_opt32;
+/// MurmurHash3_128_opt64 $(D std.digest.digest) Template API.
+alias MurmurHash3_128_opt64 = Piecewise!FastMurmurHash3_128_opt64;
+
+/// MurmurHash3_32_opt32 $(D std.digest.digest.Digest) OOO API.
+alias MurmurHash3_32_opt32Digest = WrapperDigest!MurmurHash3_32_opt32;
+/// MurmurHash3_128_opt32 $(D std.digest.digest.Digest) OOO API.
+alias MurmurHash3_128_opt32Digest = WrapperDigest!MurmurHash3_128_opt32;
+/// MurmurHash3_128_opt64 $(D std.digest.digest.Digest) OOO API.
+alias MurmurHash3_128_opt64Digest = WrapperDigest!MurmurHash3_128_opt64;
+
+/*
+Performance notes:
+ - To help a bit with the performance when compiling with DMD some functions in
+   this module have been duplicated, some other functions have been rewritten to
+   pass by value instead of by reference.
+ - GDC and LDC are on par with their C++ counterpart.
+ - DMD is typically between 20% to 50% of the GCC version.
+ -
+*/
+
+/++
+MurmurHash3 optimized for x86 processors producing a 32-bit value.
+
+This is a lower level implementation that makes finalization optional and have
+better performance than $(D digest).
+Note that $(D putRemainder) can be called only once and that no subsequent calls
+to $(D putBlocks) is allowed.
++/
+struct FastMurmurHash3_32_opt32
+{
+private:
+    enum uint c1 = 0xcc9e2d51;
+    enum uint c2 = 0x1b873593;
+    uint h1;
+
+public:
+    alias Block = uint; /// The element type for 32-bit implementation.
```

Comment:
I think the name `Block` is really confusing. I've wondered what that is all through the code review. Would `Unit` fare better?